pkg/commands/componentarchive/signature/verify/verify.go

```diff
@@ -0,0 +1,299 @@
+		}
+
+		digest, err := recursivelyCheckCds(childCd, repoContext, ociClient, ctx, hasherForCdReference, skipAccessTypes)
+		if err != nil {
+			return fmt.Errorf("checking of component reference %s:%s failed: %w", reference.ComponentName, reference.Version, err)
+		}
+
+		if !reflect.DeepEqual(reference.Digest, digest) {
+			return fmt.Errorf("component reference digest for  %s:%s is different to stored one", reference.ComponentName, reference.Version)
+		}
+
+	}
+	for _, resource := range cd.Resources {
+		log := logger.Log.WithValues("componentDescriptor", cd, "resource.name", resource.Name, "resource.version", resource.Version, "resource.extraIdentity", resource.ExtraIdentity)
+
+		//skip ignored access type
+		if _, ok := skipAccessTypesMap[resource.Access.Type]; ok {
+			log.Info("skipping resource as defined in --skip-access-types")
+			continue
+		}
+		if resource.Digest == nil || resource.Digest.HashAlgorithm == "" || resource.Digest.NormalisationAlgorithm == "" || resource.Digest.Value == "" {
+			return fmt.Errorf("resource is missing digest %s:%s", resource.Name, resource.Version)
+		}
+
+		hasher, err := cdv2Sign.HasherForName(resource.Digest.HashAlgorithm)
+		if err != nil {
+			return fmt.Errorf("failed creating hasher for algorithm %s for resource %s %s: %w", resource.Digest.HashAlgorithm, resource.Name, resource.Version, err)
+		}
+		digester := signatures.NewDigester(ociClient, *hasher, skipAccessTypes)
+
+		digest, err := digester.DigestForResource(ctx, *cd, resource)
+		if err != nil {
+			return fmt.Errorf("failed creating digest for resource %s: %w", resource.Name, err)
+		}
+
+		if !reflect.DeepEqual(resource.Digest, digest) {
+			return fmt.Errorf("resource digest is different to stored one %s:%s", resource.Name, resource.Version)
+		}
+
+	}
+	return nil
+}
+
+func recursivelyCheckCds(cd *cdv2.ComponentDescriptor, repoContext cdv2.OCIRegistryRepository, ociClient ociclient.Client, ctx context.Context, hasherForCd *cdv2Sign.Hasher, skipAccessTypes []string) (*cdv2.DigestSpec, error) {
+	skipAccessTypesMap := map[string]bool{}
+	for _, v := range skipAccessTypes {
+		skipAccessTypesMap[v] = true
+	}
+
+	for referenceIndex, reference := range cd.ComponentReferences {
+		reference := reference
+
+		ociRef, err := cdoci.OCIRef(repoContext, reference.Name, reference.Version)
+		if err != nil {
+			return nil, fmt.Errorf("invalid component reference: %w", err)
+		}
+
+		cdresolver := cdoci.NewResolver(ociClient)
+		childCd, err := cdresolver.Resolve(ctx, &repoContext, reference.ComponentName, reference.Version)
+		if err != nil {
+			return nil, fmt.Errorf("unable to to fetch component descriptor %s: %w", ociRef, err)
+		}
+
+		var hasher *cdv2Sign.Hasher
+		//default to sha256 as default
+		if reference.Digest == nil || reference.Digest.HashAlgorithm == "" {
+			var err error
```

Comment:
Get hash algorithm id in the "if" condition, then handle object creation outside of if condition only once.